server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java

```diff
@@ -666,16 +677,28 @@ Releasable tryAcquireSnapshotDownloadPermits() {
         final int maxConcurrentSnapshotFileDownloads = getMaxConcurrentSnapshotFileDownloads();
         final boolean permitAcquired = maxSnapshotFileDownloadsPerNodeSemaphore.tryAcquire(maxConcurrentSnapshotFileDownloads);
         if (permitAcquired == false) {
-            logger.warn(
-                String.format(
-                    Locale.ROOT,
-                    "Unable to acquire permit to use snapshot files during recovery, "
-                        + "this recovery will recover index files from the source node. "
-                        + "Ensure snapshot files can be used during recovery by setting [%s] to be no greater than [%d]",
-                    INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS.getKey(),
-                    this.maxConcurrentSnapshotFileDownloadsPerNode
-                )
-            );
+            if (this.maxConcurrentIncomingRecoveries <= this.maxConcurrentSnapshotFileDownloadsPerNode) {
+                logger.warn(
+                    String.format(
+                        Locale.ROOT,
+                        "Unable to acquire permit to use snapshot files during recovery, "
+                            + "this recovery will recover index files from the source node. "
+                            + "Ensure snapshot files can be used during recovery by setting [%s] to be no greater than [%d]",
+                        INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS.getKey(),
+                        this.maxConcurrentSnapshotFileDownloadsPerNode / Math.max(1, this.maxConcurrentIncomingRecoveries)
+                    )
+                );
+            } else {
+                logger.warn(
+                    String.format(
+                        Locale.ROOT,
+                        "Number of concurrent incoming recoveries exceeds max number of concurrent snapshot file downloads "
+                            + "per node. Thus some recoveries won't be able to make a progress. Consider decreasing value of "
+                            + "[cluster.routing.allocation.node_concurrent_incoming_recoveries] or disabling use of "
+                            + "snapshot recovery entirely by setting [indices.recovery.use_snapshots] to [false]"
```

Comment:
I'd prefer we use `CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_INCOMING_RECOVERIES_SETTING.getKey()` and `INDICES_RECOVERY_USE_SNAPSHOTS_SETTING.getKey()` rather than the literal setting names in the message. Not that they'll ever change, but it makes it possible to find references to the settings in your IDE.

Also could we include the current values of `maxConcurrentIncomingRecoveries` and `this.maxConcurrentSnapshotFileDownloadsPerNode` in the message?

Also (nit) perhaps a multi-line string would be preferable as the format string here?

Reply:
Done